Hyperbolic_triangulation_2/include/CGAL/Hyperbolic_Delaunay_triangulation_2.h

```diff
@@ -0,0 +1,956 @@
+
+      // Cannot be on the boundary here.
+      lt = FACE;
+      if (cs1 != cp1 || cs2 != cp2 || cs3 != cp3) {
+        return ON_NEGATIVE_SIDE;
+      } else {
+        return ON_POSITIVE_SIDE; 
+      }
+  }
+
+
+  int get_finite_non_hyperbolic_edge(Face_handle f) const
+  {
+    CGAL_triangulation_precondition(is_finite_non_hyperbolic(f));
+    Face_data fd = object_cast<Face_data>(f->tds_data());
+    return fd.get_non_hyperbolic_edge(); 
+  }
+  
+  bool is_finite_non_hyperbolic(Face_handle f) const
+  {
+    if (const Face_data* td = object_cast<Face_data>(&f->tds_data())) {
+      Face_data fd = object_cast<Face_data>(f->tds_data());
+      return !fd.get_is_Delaunay_hyperbolic();
+    } else {
+      return false;
+    }
+  }
+  
+  bool is_finite_non_hyperbolic(Face_handle f, int i) const
+  {
+    if(this->dimension() <= 1) {
+      return false;
+    }
+    
+    if(is_finite_non_hyperbolic(f) && get_finite_non_hyperbolic_edge(f) == i) {
+      return true;
+    }
+    
+    // another incident face and corresponding index
+    Face_handle f2 = f->neighbor(i);
+    int i2 =  f2->index(f);
+    
+    if(is_finite_non_hyperbolic(f2) && get_finite_non_hyperbolic_edge(f2) == i2) {
+      return true;
+    }
+    
+    return false;
+  }
+  
+  bool is_finite_non_hyperbolic(const Edge& e) const
+  {
+    return is_finite_non_hyperbolic(e.first, e.second);
+  }
+  
+  // Depth-first search (dfs) and marking the finite non_hyperbolic faces.
+  void mark_finite_non_hyperbolic_faces() const
+  {
+    if(this->dimension() <= 1) return;
+      
+    std::set<Face_handle> visited_faces;
+    
+    // maintain a stack to be able to backtrack
+    // to the most recent faces which neighbors are not visited
+    std::stack<Face_handle> backtrack;
+    
+    // start from a face with infinite vertex
+    Face_handle current = Base::infinite_face();
+    
+    // mark it as visited
+    visited_faces.insert(current);
+    
+    // put the element whose neighbors we are going to explore.
+    backtrack.push(current);
+    
+    Face_handle next;
+    
+    while(!backtrack.empty()) {
+      // take a face
+      current = backtrack.top();
+      
+      // start visiting the neighbors
+      int i = 0;
+      for(; i < 3; i++) {
+        next = current->neighbor(i);
+        
+        // if a neighbor is already visited, then stop going deeper
+        if(visited_faces.find(next) != visited_faces.end()) {
+          continue;
+        }
+        
+        visited_faces.insert(next);
+        mark_face(next);
+
+        // go deeper if the neighbor is non_hyperbolic
+        if(!is_Delaunay_hyperbolic(next)) {
+          backtrack.push(next);
+          break;
+        }
+      }
+      
+      // if all the neighbors are already visited, then remove "current" face.
+      if(i == 3) {
+        backtrack.pop();
+      }
+    }
+    
+  }
+  
+  // check if a star is bounded by finite faces
+  // TODO: rename this function name
```

Comment:
I have implemented most of the missing functions. The only things that remain as \todo are the `operator==` and `operator!=` functions in both `HT2` and `P4HT2`, but I wonder if they are really needed... 

I have found these operators implemented for `Triangulation_3`, `Periodic_3_triangulation_3`, and `Periodic_2_triangulation_2`, so it might make sense to have them for `HT2` and `P4HT2`, but I want to make sure that they're needed before implementing them, as it will take me some time, from what I saw in the existing methods.